### PR TITLE
YAML configuration file

### LINKED ISSUE #133: Need check for model definition in configuration file
A construction of the TOML file like following makes FAST crash 
```
    [model.aerodynamics]
        id = "something"
    [model.aerodynamics.landing]
        id = "something.else"
```

Such construction should raise an error message explaining it is wrong. Definition of `model.aerodynamics.landing` makes that `model.aerodynamics` is a group, but it is also defined as a component because it has an ID.

### LINKED ISSUE #270: Current format for configuration file could be enhanced
FAST-OAD usage has shown that the current format for configuration file could maybe be enhanced.
As some other configuration information will come, in this file or another (beginning with the descriptions of sizing and operational missions), it is probably a good time for thinking this format again.

To begin with, here is a sample file:
```toml
title = "Current FAST-OAD format"

# List of folder paths where user added custom registered OpenMDAO components
module_folders = ["my/folder", "some/other/folder"]

# Input and output files
input_file = "./problem_inputs.xml"
output_file = "./problem_outputs.xml"

# Definition of problem driver assuming the OpenMDAO convention "import openmdao.api as om"
driver = "om.ScipyOptimizeDriver(tol=1e-6, optimizer='COBYLA')"


# Definition of OpenMDAO model
[model]
# Solvers are defined assuming the OpenMDAO convention "import openmdao.api as om"
    nonlinear_solver = "om.NonlinearBlockGS(maxiter=200)"
    linear_solver = "om.DirectSolver()"

    # Although "model" is a mandatory name for the top level of the model, its
    # sub-components can be freely named by user
    [model.my_first_model]
        # An OpenMDAO component is identified by its "id"
        id = "my.first.model"
    [model.my_group]
        [model.my_group.my_second_model]
            id = "my.second.model"
        [model.my_group.my_third_model]
            id = "my.third.model"
            some_model_option = 11



[optimization]  # Actually, this line can be omitted
# This section is needed only if optimization process is run

    # Double brackets are needed when using multiple occurences of the same key
    [[optimization.design_var]]
        name = "first:design_var"
        lower = 0.0
        upper = 42.0
    [[optimization.design_var]]
        name = "second:design_var"
        lower = -100.0
        upper = 0.0

    [[optimization.constraint]]
        name = "first:constraint"
        lower = 0.05
        upper = 0.1
    [[optimization.constraint]]
        name = "second:constraint"
        lower = -1.5
        upper = 1.0   # TOML format does not accept "1."

    [[optimization.objective]]
        name = "one:objective:function"
```

Here are some problems I saw in relation with current formatting (please feel free:
1. There is a mix of keys that can be named by user (model names, group names) and keys that have fixed names ("model", "design_var", ...)
1. There is a mix of simple and double brackets that is part of TOML syntax, but that can be confusing for users.
1. Nested structures require to repeat the name of parent structure (e.g. **`model`**`.my_first_model`)
1. TOML format does not support floats that end with a dot: `1.0` is Ok, but `1.` is not.

Some modifications of current format could be implemented to fix some of these problems. For instance, even if it is not required by TOML syntax, we could decide that double-bracketed keys have fixed name, whereas simple-bracketed keys are free to name. Also, as design variables, constraints and objective are obviously related to optimization, the "optimization" key might be removed.
That would give:
```toml
title = "Possible alternative to current format"

# List of folder paths where user added custom registered OpenMDAO components
module_folders = ["my/folder", "some/other/folder"]

# Input and output files
input_file = "./problem_inputs.xml"
output_file = "./problem_outputs.xml"

# Definition of problem driver assuming the OpenMDAO convention "import openmdao.api as om"
driver = "om.ScipyOptimizeDriver(tol=1e-6, optimizer='COBYLA')"


# Definition of OpenMDAO model
[[model]]
# Solvers are defined assuming the OpenMDAO convention "import openmdao.api as om"
    nonlinear_solver = "om.NonlinearBlockGS(maxiter=200)"
    linear_solver = "om.DirectSolver()"

    # Although "model" is a mandatory name for the top level of the model, its
    # sub-components can be freely named by user
    [model.my_first_model]
        # An OpenMDAO component is identified by its "id"
        id = "my.first.model"
    [model.my_group]
        [model.my_group.my_second_model]
            id = "my.second.model"
        [model.my_group.my_third_model]
            id = "my.third.model"
            some_model_option = 11


# Optimization settings
[[design_var]]
    name = "first:design_var"
    lower = 0.0
    upper = 42.0
[[design_var]]
    name = "second:design_var"
    lower = -100.0
    upper = 0.0

[[constraint]]
    name = "first:constraint"
    lower = 0.05
    upper = 0.1
[[constraint]]
    name = "second:constraint"
    lower = -1.5
    upper = 1.0

[[objective]]
    name = "one:objective:function"
```

Another way might be to change to YAML, which is also a popular format. Unlike TOML, indentation is part of YAML syntax. A sample file could be:
```yaml
title: YAML alternative
module_folders: [ my/folder, some/other/folder ]

# Input and output files
input_file: ./problem_inputs.xml
output_file: ./problem_outputs.xml

# Definition of problem driver assuming the OpenMDAO convention import openmdao.api as om
driver: om.ScipyOptimizeDriver(tol:1e-6, optimizer:'COBYLA')

# Definition of OpenMDAO model
model_settings:
  # Solvers are defined assuming the OpenMDAO convention import openmdao.api as om
  nonlinear_solver: om.NonlinearBlockGS(maxiter:200)
  linear_solver: om.DirectSolver()

model:
  # Although "model" is a mandatory name for the top level of the model, its
  # sub-components can be freely named by user
  my_first_model:
    id: my.first.model
  my_group:
    my_second_model:
      id: my.second.model
    my_third_model:
      id: my.third.model
      some_model_option: 11


# This section is needed only if optimization process is run
optimization:
  design_vars:
    - name: first:design_var
      lower: 0.0
      upper: 42.0
    - name: second:design_var
      lower: -100.0
      upper: 0.0

  constraints:
    - name: first:constraint
      lower: 0.05
      upper: 0.1
    - name: second:constraint
      lower: -1.5
      upper: 1.

  objective: one:objective:function
```

I like some features of YAML:
- indentation decides what is nested in what --> no need to repeat parent name
- strings do not need to be quoted
- `1.` is accepted

On the other side, I fear that some user may break indentations by accidentally adding/removing some space. (although such problems may be identified at runtime using schema validation...)

What do you think? Any opinion is welcome !

_Originally posted by @christophe-david in https://github.com/fast-aircraft-design/FAST-OAD/discussions/213_

### PR DESCRIPTION
**Pull request by christophe-david**
This PR resolves #270.

With this PR, YAML is now the default format for FAST-OAD configuration files. TOML format is still accepted, though.

Documentation has been updated and can be checked [here](https://fast-oad.readthedocs.io/en/issue-270_yaml-configuration-file/documentation/usage.html#fast-oad-configuration-file).

I have hesitated a lot about the yaml parser, between `strictYAML`, `pyyaml` and `ruamel.yaml`:
- I found `strictYAML` interesting, because it aims at not implementing the YAML features that can make YAML unsafe or hard to read. Unlike the 2 others, that need validation with `jsonschema` after reading, it can do schema validation while reading. That's the used framework currently used for parsing mission definition files. But its schema has to be implemented in Python, which is much more powerful than a classic JSON schema, except that JSON schemas allow autocompletion in most IDEs, which is a feature we will really need (and even more for mission definition files! So we will probably change the parser for mission definition files soon)
- `pyyaml` is certainly the most used around the world, but it currently supports only YAML 1.1. And in YAML 1.1, `1e-4` is not recognized as a number, whereas `1.e-4` is. And I find it annoying.
- `ruamel.yaml` is an active fork of `pyyaml` that supports YAML 1.2, so it is the winner :)

A JSON schema has been defined, which resolves #133.

____

I also changed some keywords:
- `connection` -> `connections`
- `design_var` -> `design_variables`
- `constraint` -> `constraints`

While these keywords had to be repeated in TOML format, they are unique in YAML, so the plural is more intuitive. It breaks the backward compatibility with old TOML files, but I think it is worth it.

### CODE REVIEW

**Review by ScottDelbecq — CHANGES_REQUESTED** [review]
Thank you for this PR. I think moving to YAML for the configuration file is a good decision.
I spotted a few things in the docs (cf. my comments).
Also, the text in the tutorial notebook with `.toml` should be replaced by `.yml`. 

By the way, I wonder how there is no conflict for merging this branch into the master since the wing position MDA branch has been merged (#268), leading to a different optimization problem definition than in this branch?

File: docs/documentation/usage.rst [hunk 1]
```diff
@@ -16,204 +16,236 @@ OpenMDAO systems. This part is addressed in the :ref:`API documentation<fastoad>
...
 Custom module path
 ==================
 
-.. code:: toml
+.. code:: yaml
 
-    module_folders = []
+    module_folders:
 
 Provides the path where user can have his custom OpenMDAO modules. See section :ref:`add-modules`.
```
[the remark this answers is not in the document]
Comment (christophe-david) [reply docs/documentation/usage.rst:100]: Well seen. Done in f255b7d705ef623729f451797c03ac369bc7b386.

**Comment by christophe-david** [comment]
> Thank you for this PR. I think moving to YAML for the configuration file is a good decision.
> I spotted a few things in the docs (cf. my comments).

Just to be sure, I see only one comment, but you use plural. Is it normal ?

> Also, the text in the tutorial notebook with `.toml` should be replaced by `.yml`.

Done in 75d2d8f6860ea43ae2efc3885a64ceae280d799e. BTW, I cleared the outputs of the notebooks. I confirm what I said earlier, that I cannot see any useful output from previous runs when I open a notebook.

> By the way, I wonder how there is no conflict for merging this branch into the master since the wing position MDA branch has been merged (#268), leading to a different optimization problem definition than in this branch?

Simply because after the merge of #268, I updated this branch by merging master into it (commit 03491e643dc1453cd73c4f32a9b1a581021711bd). I just forgot to search again the notebooks for references to TOML 😅.

Thank you for the review, I let you check again if all is OK.

**Comment by ScottDelbecq** [comment]
Nope, my bad for the plural!
Ok thanks then :)

**Review by ScottDelbecq — APPROVED** [review]

**Pull request merged**